### ISSUE TITLE
PCA error on own dataset

### ISSUE THREAD

**Issue by juliangaal** [facebookresearch/deepcluster#40]
I'm running into this error with my own dataset:
```bash
RuntimeError: Error in void faiss::PCAMatrix::prepare_Ab() at VectorTransform.cpp:482: Error: 'd_out * d_in <= PCAMat.size()' failed: PCA matrix cannot output 256 dimensions from 4096
```
My environment setup matches your defined dependencies (except cuda10, which may become an issue...?).

Parameters I tested, but resulted in the same errror:
* alexnet vs vgg16
* kmeans vs PIC
* images preprocessed manually to 256x256 and 3channel grayscale (original images grayscale) vs preprocessing handled by [your code](https://github.com/facebookresearch/deepcluster/blob/933a3ed3ee9f8c618c565def37861df1f1617b6a/main.py#L130)

Thanks for your help!

**Comment by mathildecaron31** [comment]
Hi, thanks for your interest !
Looking at the error it seems that you don't have enough data points in your dataset.
Indeed I manage to reproduce your error when I use less than 256 points.
Have a look at faiss pca [code](https://github.com/facebookresearch/faiss/blob/656368b5eda4d376177a3355673d217fa95000b6/VectorTransform.cpp#L480):
if `n` the number of points is higher or equal than the points dimension, `d_in`, then `PCAMat is of dim `d_in * d_in` and the [condition]((https://github.com/facebookresearch/faiss/blob/656368b5eda4d376177a3355673d217fa95000b6/VectorTransform.cpp#L480) is True.
However, if `n < d_in` (which I assume is your case) then `n` needs to be higher than d_out for the condition to be True.

In a nutshell, it will throw the error if `n < d_in`, in your case it seems that your dataset size is less than 256.

This seems to be a pure faiss error, so if I am not able to solve your issue you might post on [faiss](https://github.com/facebookresearch/faiss) directly.

Hope it helps

**Comment by juliangaal** [comment]
Ah, great to hear, thanks a lot. Yes, my first test set is VERY small to just see if the whole pipeline works. Cheers

**Comment by skasapis** [comment]
Hey, I had the same problem with a dataset of 100 pictures. In the code I am using the function sets the pca variable to be 256. You can drop this to lets say 50, and you will solve your problem as I did! 

def preprocess_features(npdata, pca=256): #change this to 50
mat = faiss.PCAMatrix (ndim, pca, eigen_power=-0.5)
mat.train(npdata)

**Comment by BEFOURD** [comment]
you must let the number of your dataset n % batch_size == 0 or n % batch_size >= d_out
